Initial check-in of the Shim layer.

carlosalberto opened the pull request:
This is an updated PR for putting the Shim layer in a separated repo (originated from opentracing/opentracing-java#221, and tuned as per Yuri's suggestion to only ask for a review of the very Shim layer).

Notes on it overall:

* Provides a default TracerShim which wraps a 0.31 Tracer with its ScopeManager, providing no Continuation support out of the box.

* Provides also a AutoFinishTracerShim which wraps a 0.31 Tracer using the contained AutoFinishScopeManager, which keeps the thread-local, reference-count based logic to support Continuations.

* It still throws IllegalArgumentException on null values of some methods (i.e. SpanBuilder.childOf() - throwing this would prevent a NullReferenceException when casting the wrapped object. Pavol mention we should not throw such Exception - so, what should we throw there instead? Or how should we report such errors? The spec doesn't mention thrown exceptions...

* Currently it's a minimalistic approach, and wondering if we this is good enough for `Tracer` authors who were using their own span propagation and/or using their own Span lifetime handling.

No Travis/CentralSync support yet (will do after we get this properly merged).

carlosalberto committed:
Initial check-in of the Shim layer.

carlosalberto commented:
Updated PR with only the Shim classes + AutoFinishScopeManager. @pavolloffay @malafeev @yurishkuro. Let me know ;)

yurishkuro commented:
Haven't gone through it line by line, but read through a few classes, lgtm.  Is anything in particular of a concern?  Did we try adapting existing instrumentations to use this?

carlosalberto committed:
Base ActiveSpanShim/SpanShim.equals() on the wrapped value.

This way we can correctly compare two ActiveSpanShim/SpanShim
objects pointing to the very same Scope/Span object,
respectively.

carlosalberto commented:
So far only with `MockTracer`. I *think* @malafeev wanted to try it out with the Lightstep tracer ;)

malafeev commented:
Yes, I just tested it with LightStep tracer (upgraded to RC1) mixing two instrumentations: JMS with compatibility layer from this PR
and RxJava with 0.30. It works for me pretty well.

carlosalberto commented:
Merging. I will enable bintray/maven/travis integration and will roll out a release candidate soon. Thanks for the feedback.
